### PR TITLE
Add stream support in storage clients (Closes #1702)

### LINKED ISSUE #1702: Extend storage clients to handle Streams when used in node
Resolving #1690 in a manageable way requires that we divide the changes needed into separate independent PRs. First, the storage clients should be able to operate with streams in Node.

Goals to be resolved in this issue:

-  [x]  Extend the `BlobClient` to `StreamBlobClient` for supporting streams. This can be achieved relatively straight forward manner since streams are supported in `node-fetch` and `superagent`.

-  [x] Change `StorageBackend` and `Clients` with support for `getStream` and `putStream`. One question is should the existing `getFile` and `putFile` methods be updated or should we create separate methods to support streams?
- [x] Update tests for StorageClients.

### PR DESCRIPTION
**Pull request by umesh-timalsina**
From Commit Logs:

```
    1.  Extend BlobClient to support streams
    2.  Change StorageBackend clients with getStream and putStream methods
    3.  Update tests for streams in StorageBackends.spec.js
    4.  Update aws-sdk and remove pormisify in S3Client
```

### CODE REVIEW

File: src/common/storage/backends/s3/Client.js [hunk 1]
```diff
@@ -110,20 +107,44 @@ define([
...
+    };
+
     S3Storage.prototype.deleteDir = async function (dirname) {
         const s3Client = await this.getS3Client();
         const {Contents} = await s3Client.listObjectsV2({
             Bucket: this.bucketName,
             MaxKeys: 1000,
             Prefix: dirname
-        });
+        }).promise();
```
Comment (brollb) [inline src/common/storage/backends/s3/Client.js:140]: Switching to `.promise` rather than the `promisifyMethod` is a nice refactor. It would be a little cleaner to move this type of refactor to its own PR.

File: package.json [hunk 2]
```diff
@@ -28,7 +28,7 @@
     "@babel/polyfill": "^7.7.0",
     "@babel/preset-env": "^7.7.1",
     "@babel/runtime": "^7.7.2",
-    "aws-sdk": "^2.616.0",
+    "aws-sdk": "^2.678.0",
```
Comment (brollb) [inline package.json:31]: I assume this is part of the refactor to use `.promise()`, right?

File: src/common/StreamBlobClient.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    'blob/BlobClient',
+    'deepforge/StorageHelpers'
+], function(
+    Logger,
+    gmeConfig,
+    BlobClient,
+    StorageHelpers
+) {
+    const StreamBlobClient = function (logger) {
+        const params = this.getBlobClientParams(logger);
```
Comment (brollb) [inline src/common/StreamBlobClient.js:14]: As this is subclassing `BlobClient` and the hope is to contribute this back to webgme-engine, this should not modify the method signature unnecessarily. The creation of the blob client parameters should be handled by the storage adapter itself.

File: src/common/storage/backends/StorageClient.js [hunk 4]
```diff
@@ -1,14 +1,13 @@
 /* globals define*/
 define([
     'client/logger',
-    'deepforge/gmeConfig'
+    'deepforge/gmeConfig',
+    'deepforge/StorageHelpers'
```
Comment (brollb) [inline src/common/storage/backends/StorageClient.js:5]: If this is going to be refactored, `StorageHelpers` should be in `deepforge/storage`.

File: src/common/storage/backends/StorageClient.js [hunk 5]
```diff
@@ -87,5 +70,11 @@ define([
         throw new Error(`stat not implemented for ${this.name}`);
     };
 
+    StorageClient.prototype.checkStreamsInBrowser = async function() {
```
Comment (brollb) [inline src/common/storage/backends/StorageClient.js:73]: Personally, I would rename this to something like `ensureStreamsSupported` or `ensureIsBrowser(msg)`. This doesn't actually check streams and it appears that streaming in the browser is partially supported (`getStream` appears to be supported).

Comment (brollb) [reply src/common/storage/backends/StorageClient.js:73]: Also, this shouldn't be `async`

File: test/integration/StorageBackends.spec.js [hunk 6]
```diff
@@ -75,4 +96,16 @@ describe('Storage Features Test', function () {
         }
         return 1;
     }
+
+    async function verifyStreamContent(inputStream) {
```
Comment (brollb) [inline test/integration/StorageBackends.spec.js:100]: This should be possible without writing to a file. Can't you just collect it in memory and check it?

File: test/unit/common/storage/backends/gme/Client.spec.js [hunk 7]
```diff
@@ -1,4 +1,4 @@
-describe('GME Storage Adapter', function() {
+describe.skip('GME Storage Adapter', function() {
```
Comment (brollb) [inline test/unit/common/storage/backends/gme/Client.spec.js:1]: Why are these skipped?

**Comment by brollb** [comment]
I just merged #1710 into master. Could you merge master in (and resolve the merge conflicts)?

**Comment by umesh-timalsina** [comment]
Sure.

**Comment by umesh-timalsina** [comment]
This now depends on https://github.com/webgme/webgme-engine/pull/215/

File: test/integration/StorageBackends.spec.js (hunk 6 above)
Comment (umesh-timalsina) [reply test/integration/StorageBackends.spec.js:102]: This method now uses a `StringWritable` pipeline.

**Comment by brollb** [comment]
This is looking good. Could you check out the source of the failing tests? Are they an issue with sciserver or with our request to sciserver?
DeepinScreenshot_select-area_20200721113407

**Comment by brollb** [comment]
Main concerning failing tests:
DeepinScreenshot_select-area_20200722141945

**Comment by umesh-timalsina** [comment]
These tests are a passing for me. 
image
